### PR TITLE
Remove deprecated torch.solve

### PR DESCRIPTION
**Pull request by IvanYashchuk**
The time has come to remove deprecated linear algebra related functions. This PR removes `torch.solve`.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

### CODE REVIEW

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -278,22 +270,6 @@ template<class scalar_t, class value_t=scalar_t>
...
-template<> void lapackSolve<c10::complex<float>>(int n, int nrhs, c10::complex<float> *a, int lda, int *ipiv, c10::complex<float> *b, int ldb, int *info) {
-  cgesv_(&n, &nrhs, reinterpret_cast<std::complex<float>*>(a), &lda, ipiv, reinterpret_cast<std::complex<float>*>(b), &ldb, info);
-}
-
-template<> void lapackSolve<double>(int n, int nrhs, double *a, int lda, int *ipiv, double *b, int ldb, int *info) {
-  dgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, info);
-}
-
-template<> void lapackSolve<float>(int n, int nrhs, float *a, int lda, int *ipiv, float *b, int ldb, int *info) {
-  sgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, info);
```
Comment (Lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:294]: Why don't we want to keep these kernels? Shouldn't they be more efficient than lu + lu_solve?

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -163,86 +153,6 @@ void magmaGels(
...
     scalar_t* hwork, magma_int_t lwork, magma_int_t* info);
 
-template<>
-void magmaSolve<double>(
-    magma_int_t n, magma_int_t nrhs, double* dA, magma_int_t ldda,
-    magma_int_t* ipiv, double* dB, magma_int_t lddb, magma_int_t* info) {
-  MagmaStreamSyncGuard guard;
-  magma_dgesv_gpu(n, nrhs, dA, ldda, ipiv, dB, lddb, info);
-  AT_CUDA_CHECK(cudaGetLastError());
-}
```
Comment (Lezcano) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:173]: Same question for these.

**Review by lezcano — COMMENTED** [review]
I won't cry when this one's gone either. Now, I left a few comments regarding the MAGMA / LAPACK functions that we're removing.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:294]: No, they are not more efficient. `gesv` = `getrf` + `getrs` but done in an external library. We should leave only the code that is used.

**Comment by mruberry** [comment]
Build issues look real here, too

**Comment by IvanYashchuk** [comment]
I resolved the build issues.

**Review by lezcano — APPROVED** [review]
Happy with this one once the CI is green.

I will resolve the merge conflicts in https://github.com/pytorch/pytorch/pull/74046 when this gets merged. cba to roll them back now really.

**Review by albanD — COMMENTED** [review]
Thanks for the PR Ivan.
Do you think it would be possible to add to `torch/__init__.py` something like:
```python
# This function was depreacted and removed.
# This nice error message can be removed in version 1.3+
def solve(input, A, out=None):
    raise RuntimeError("This function was deprecated since version 1.9 and is now removed. Please use the `torch.linalg.solve` function instead")
```

To make sure that users get a nice error message for a couple versions?

**Comment by IvanYashchuk** [comment]
Sure, I will add that.

**Comment by IvanYashchuk** [comment]
Alright, I added a `solve` function and a method that raises an error:
```py
In [1]: import torch
In [2]: a = torch.randn(3, 3)
In [3]: b = torch.randn(3, 3)
In [4]: torch.solve(b, a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-6f9b9974b5b1> in <module>
----> 1 torch.solve(b, a)

RuntimeError: This function was deprecated since version 1.9 and is now removed. Please use the `torch.linalg.solve` function instead.

In [5]: b.solve(a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-fd669f2275c3> in <module>
----> 1 b.solve(a)

RuntimeError: This function was deprecated since version 1.9 and is now removed. Please use the `torch.linalg.solve` function instead.
```

**Comment by Lezcano** [comment]
If you ever manage to land this, remind me to remove in https://github.com/pytorch/pytorch/pull/74046 the backward functions for `torch.solve`. Big if though.

File: test/test_linalg.py [hunk 3]
```diff
@@ -6351,12 +6264,10 @@ def test_solve_methods_arg_device(self, device):
             b = torch.randn(3, 1, device=b_device)
             A = torch.randn(3, 3, device=A_device)
 
-            # solve and cholesky_solve goes through generic backend dispatch and hit kernel specific device check first
+            # cholesky_solve goes through generic backend dispatch and hit kernel specific device check first
             # triangular_solve goes through specific backend dispatch (CPU/CUDA) and hit auto-generated device check first
             generic_backend_dispatch_err_str = "Expected b and A to be on the same device"
             specific_backend_dispatch_err_str = "Expected all tensors to be on the same device"
-            with self.assertRaisesRegex(RuntimeError, generic_backend_dispatch_err_str):
```
Comment (albanD) [inline test/test_linalg.py:6358]: nit: could you leave a small test here (or in test_torch.py) that the error you added is working as expected?

**Review by albanD — COMMENTED** [review]
Sounds good!

File: test/test_linalg.py (hunk 3 above)
Comment (IvanYashchuk) [reply test/test_linalg.py:6367]: Sure, I added the test in https://github.com/pytorch/pytorch/pull/70986/commits/68928e736f23e2d922023ac84f5f2622a4e338ed.

**Comment by IvanYashchuk** [comment]
I added a test checking that the error is raised as expected and resolved accumulated merge conflicts.

**Review by albanD — APPROVED** [review]
Sounds good!

**Pull request closed without merging**